### PR TITLE
Add get_min_timestamp to Bundle

### PR DESCRIPTION
**Pull request by nedgar**
Improves performance by taking timestamps from _StackedWindowedValues
rather than iterating over all elements.
Memoizes min timestamp if bundle is committed.

### CODE REVIEW

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by nedgar** [comment]
See thread at https://twitter.com/nedgar/status/929765848750977026
Basically, the min timestamp calculation in `_TransformWatermarks.refresh()` is expensive due to iterating over all elements in the bundle, and is called repeatedly. By adding `get_min_timestamp` to Bundle, it need only iterate over the root elements, e.g. taking timestamp directly from _StackedWindowedValues rather than iterating over their elements.
It also memoizes the min timestamp if the bundle is committed.

**Comment by nedgar** [comment]
Before:
```
$ time python wordcount_minimal.py --input kinglear.txt --output counts
...
real	0m4.726s
user	0m4.457s
sys	0m0.214s

$ time python wordcount_minimal.py --input kinglear_x4.txt --output counts
...
real	0m22.268s
user	0m21.744s
sys	0m0.722s
```

After:
```
$ time python wordcount_minimal.py --input kinglear.txt --output counts
...
real	0m3.893s
user	0m3.623s
sys	0m0.189s

$ time python wordcount_minimal.py --input kinglear_x4.txt --output counts
...
real	0m10.994s
user	0m10.404s
sys	0m0.472s
```

Where `kinglear_x4.txt` is `kinglear.txt` repeated 4 times.

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by aaltay** [comment]
R: @charlesccychen

File: sdks/python/apache_beam/runners/direct/bundle_factory.py [hunk 1]
```diff
@@ -198,3 +200,16 @@ def commit(self, synchronized_processing_time):
     self._committed = True
     self._elements = tuple(self._elements)
     self._synchronized_processing_time = synchronized_processing_time
+
+  def get_min_timestamp(self):
+      """Returns the minimum timestamp across all elements in this bundle, or
+      MAX_TIMESTAMP if this bundle is empty."""
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/bundle_factory.py]: Can you make this one line?  For example:

```
   """Returns max element timestamp in bundle or MAX_TIMESTAMP if empty."""
```

File: sdks/python/apache_beam/runners/direct/watermark_manager.py [hunk 2]
```diff
@@ -211,12 +211,9 @@ def refresh(self):
...
       has_pending_elements = False
       for input_bundle in self._pending:
-        # TODO(ccy): we can have the Bundle class keep track of the minimum
-        # timestamp so we don't have to do an iteration here.
-        for wv in input_bundle.get_elements_iterable():
-          has_pending_elements = True
-          if wv.timestamp < min_pending_timestamp:
-            min_pending_timestamp = wv.timestamp
+        if input_bundle.has_elements():
+            min_pending_timestamp = min(min_pending_timestamp, input_bundle.get_min_timestamp())
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/watermark_manager.py]: Please use 2sp indentation, wrap line to 80 chars.

File: sdks/python/apache_beam/runners/direct/bundle_factory.py [hunk 3]
```diff
@@ -198,3 +200,16 @@ def commit(self, synchronized_processing_time):
     self._committed = True
     self._elements = tuple(self._elements)
     self._synchronized_processing_time = synchronized_processing_time
+
+  def get_min_timestamp(self):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/bundle_factory.py]: Please match indentation of the rest of the file (2sp).

**Comment by nedgar** [comment]
Hi @charlesccychen. Here are some fixups addressing your review comments.
Should I merge with latest from apache/beam master, or will you?

Even with these changes, it takes several seconds to process even modest sized files (e.g. a few MB). Profiling suggests much of it is due to the creation of many Timestamp objects. Is the team aware of / planning to address this? Thanks.

**Comment by charlesccychen** [comment]
Thank you!  Can you rebase to the head of master?

I am currently working on a rearchitecture of the DirectRunner to improve scalability, especially for multi-core environments (https://issues.apache.org/jira/browse/BEAM-1442).  This should be ready for use in the next few months.

**Comment by charlesccychen** [comment]
LGTM

**Comment by robertwb** [comment]
We are aware of the slow performance of the Direct runner, and hope to address it more directly. 

In the meantime, you could also try running with `pipeline(runner=apache_beam.runners.portability.fn_api_runner.FnApiRunner())` which can help on some pipelines.

**Comment by nedgar** [comment]
I've rebased, thought not sure I've done it correctly. It looks OK in GitHub Desktop:

image

Thanks for the info re performance.

**Comment by nedgar** [comment]
After the merge, running it fails due to missing `beam_runner_api_pb2`:
```
$ time python wordcount_minimal.py --input kinglear_x4.txt --output counts
Traceback (most recent call last):
  File "wordcount_minimal.py", line 53, in <module>
    import apache_beam as beam
  File "/Users/nedgar/Development/Apache-Beam/apache_beam/__init__.py", line 86, in <module>
    from apache_beam import coders
  File "/Users/nedgar/Development/Apache-Beam/apache_beam/coders/__init__.py", line 18, in <module>
    from apache_beam.coders.coders import *
  File "/Users/nedgar/Development/Apache-Beam/apache_beam/coders/coders.py", line 30, in <module>
    from apache_beam.portability.api import beam_runner_api_pb2
ImportError: cannot import name beam_runner_api_pb2
```

**Comment by kennknowles** [comment]
Any updates here? It does look like the rebase was not quite right, as a good chunk of `apache:master` appears to have been rebased as well. You might explicitly rebase like so:

  `git rebase -i HEAD~2 origin/master`

Where I have chosen `HEAD~2` since it looks like you have two commits, and `-i` because you might as well squash them together at this point.

**Comment by nedgar** [comment]
@kennknowles my apologies, I dropped the ball in getting back to this. I’m no longer exploring Beam, have switched jobs, including returning laptop with my changes. 
If you’d like to apply the proposed changes, please go ahead. 

I’m confused by the maven/gradle comments here. They seem out of context. Wrong PR or due to my bad merge?

**Pull request closed without merging**